src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -864,6 +864,32 @@ internal void SetTopLevelProperty<T>(string name, [MaybeNull] T value)
             }
         }
 
+        [DebuggerDisplay("{" + nameof(DebuggerDisplay) + ",nq}")]
+        [DataContract]
+        private class OutboundJsonRpcRequest : Protocol.JsonRpcRequest, IMessageWithTopLevelPropertyBag
+        {
+            private readonly JsonMessageFormatter formatter;
+
+            internal OutboundJsonRpcRequest(JsonMessageFormatter formatter)
+            {
+                this.formatter = formatter ?? throw new ArgumentNullException(nameof(formatter));
+            }
+
+            [JsonIgnore]
+            [IgnoreDataMember]
+            public TopLevelPropertyBag? TopLevelPropertyBag { get; set; }
+
+            public override bool TrySetTopLevelProperty<T>(string name, [MaybeNull] T value)
```

Comment:
Now that inbound/outbound is separate, do we need to implement TrySetTopLevelProperty on the inbound object?

Reply:
Maybe not. But inbound messages become outbound messages in the remote target relay case. That isn't adequately tested, I'm learning, so I didn't want to risk regressing it.

Reply:
Right, I remember that case now from unit tests. Does the relay case only work with json messages or do we support with message pack as well? I am now curious if it would work with messagepack and top level properties since we had different inbound/outbound dictionaries in message pack case.